### ISSUE TITLE
FLAC 1.3.2 appears to require an SSE3 processor (Intel x86)

### ISSUE THREAD

**Issue by sudo-rm-rf-oops** [xiph/flac#72]
I have compiled the latest FLAC sources for x86 (32 bit) and they appear to require an SSE3 processor.  Specifically, I have a report from a customer with a XEON Prestonia processor that he cannot read his existing FLAC files (at all) and this processor supports SSE2 only.

I have replicated his problem here with a very old machine (which has no SSE support at all) but I have not investigated why this is happening.  It doesn't crash - it's just that FLAC__stream_decoder_xxx functions do nothing and return FALSE.  This is all happening in a release build.  I have not checked whether debug builds are similarly affected.  I am building with Visual Studio 2017.

Is this deliberate?  I really hope not.  The version I have been using up to now (1.3.1) works fine on *all* processors, even those without any form of SSE, and I value that.  I have therefore, slightly reluctantly, reverted to that.

I did see in the change log that 1.3.2 requires SSE2 but I'm not actually sure that's true.  Certainly the code as originally written has CPU detection logic and a set of dispatch functions so I don't see why this is necessary.  I also ran a quick benchmark on my Core i3 and 1.3.1 and 1.3.2 perform identically.  Overall, this all feels like a step backwards,

**Comment by lvqcl** [comment]
Apply the patch https://github.com/xiph/flac/commit/230cffe08ff0c965de2ce111afe4513286bfda89
over the sources of FLAC 1.3.2 and test again.

**Comment by mwahh** [comment]
OK, thanks.

Why is that not in the ZIP, it was committed ages ago?  I have wasted a lot of time on this (and I don't use git).

**Comment by erikd** [comment]
Probably time for a new release.

**Comment by lvqcl** [comment]
> I already have that fix (although not at that line number), so it's not that.

It should just work then. So the problem is somewhere else.

**Comment by mwahh** [comment]
Must be.  It also happens in a debug build, so it's nothing to do with funky code generation by the compiler.  Anyway, I'm going to stick with 1.3.1 for now.  Apart from the bug I reported recently (which I have fixed in my copy) it has served me well.

**Comment by erikd** [comment]
Please try the 1.3.3 release.

**Comment by ktmf01** [comment]
I haven't been able to reproduce this with autotools-gcc. I took the source of the 1.3.2 release, compiled with gcc 10.3.0 for i686-linux-gnu and tested with `qemu-i386 -cpu 486,+cmov,+sse`, which worked fine. Using `./configure --disable-sse` I was able to run the resulting binary with `qemu-i386 -cpu 486,+cmov (which I believe is equivalent to i686) which means it should run on Pentium Pro, Pentium 2 and above, all Celerons and Xeons, AMD K6 and newer etc. Perhaps this problem is specific to Visual Studio?

@mwahh Are you still having this problem? If so, could you perhaps provide more information? You say the problem is in the decoder, but the decoder doesn't have any processor-specific optimizations. Usually programs crash when a processor gets an illegal instruction, so that makes it even more strange and harder to pinpoint.

**Comment by sudo-rm-rf-oops** [comment]
Hi ktm,

Thanks for following up on this.  It is now actually ancient history for me as I reverted to version 1.3.1 back in 2018 and my issues went away, but here's what I know about the way the problem manifested itself:

  - my code was built with Visual Studio 2017, "No enhanced instructions" (/arch:IA32), "Favour small code" (/Os)

  - the crash was indeed due to an illegal instruction and occurred when decoding FLAC files (and probably also when encoding, I didn't get that far)

  - the problem occurred in both Debug and Release builds

  - the customer in question had twin Xeon Prestonia CPUs

I also reproduced the problem on a (real) Pentium II.  I still have that machine and will look into this when I can.

One thing worth mentioning (about Visual Studio 2017).  If you compile with "Favour fast code" (/Ot) then, in Release builds, it sometimes generates SSE instructions for things like memset, even with "No enhanced instructions" set.  But I'm pretty sure that was not the issue here. IIRC, I checked that at the time.

I just changed my username BTW (to sudo-rm-rf-oops).  Didn't like the old one.

**Comment by sudo-rm-rf-oops** [comment]
@ktmf01, @erikd: Well, I finally got to the bottom of this.  It turned out to be nothing to do with the CPU type at all - it's specific to Windows XP (which I still support).

Programs built with Visual Studio 2015 and later do not correctly implement `_fstat64` when the compiled program is running on Windows XP as outlined [here](https://developercommunity.visualstudio.com/t/-stat64-windows-xp/320354).  As a result, `flac_fstat` (which maps to `_fstat64` on Windows) always fails when running on XP.  This is actually only called in one place (in `stream_decoder.c`, and (here comes the embarrassing bit) I found that I had quietly fixed it in the version of 1.3.1 that I was running and moved on.  I'm really sorry about that, I don't know what I was thinking, but that's what happened.

So, anyway, drawing a veil over that, the fix itself is very simple.  In `stream_decoder.c`, at line 3385 replace this:

```
	if(decoder->private_->file == stdin)
		return FLAC__STREAM_DECODER_LENGTH_STATUS_UNSUPPORTED;
	else if(flac_fstat(fileno(decoder->private_->file), &filestats) != 0)
		return FLAC__STREAM_DECODER_LENGTH_STATUS_ERROR;
	else {
		*stream_length = (FLAC__uint64)filestats.st_size;
		return FLAC__STREAM_DECODER_LENGTH_STATUS_OK;
	}
```

with this:

```
#ifdef _WIN32
	ifilestats.st_size = _filelengthi64 (fileno(decoder->private_->file));
	if (filestats.st_size < 0)
#else
	if(flac_fstat(fileno(decoder->private_->file), &filestats) != 0)
#endif
		return FLAC__STREAM_DECODER_LENGTH_STATUS_ERROR;
	else {
		*stream_length = (FLAC__uint64)filestats.st_size;
		return FLAC__STREAM_DECODER_LENGTH_STATUS_OK;
	}
```

Based on 1.3.3, I have attached before and after files for your convenience. `_filelength64` itself is documented [here](https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/filelength-filelengthi64?view=msvc-160).  So sorry for the noise, I'll check my facts better next time.

[xp_fix.zip](https://github.com/xiph/flac/files/7289080/xp_fix.zip)

**Comment by ktmf01** [comment]
Thanks for figuring that out, I was already in the process of compiling [Hangover](https://github.com/AndreRH/hangover) to be able to run the code through various emulated processors using QEMU+Wine, but there is no need for that now. I will double check, and also look into whether this is a problem with other compilers, like MinGW.
